example-backend/src/main/kotlin/app/truid/example/examplebackend/TruidLoginSessionFlow.kt

```diff
@@ -0,0 +1,277 @@
+        @RequestParam("error") error: String?,
+        exchange: ServerWebExchange
+    ): Void? {
+        val session = exchange.session.awaitSingle()
+
+        if (error != null) {
+            throw Forbidden(error, "There was an authorization error")
+        } else if (!verifyOauth2State(session, state)) {
+            throw Forbidden("access_denied", "State does not match the expected value")
+        } else {
+            try {
+                val body = LinkedMultiValueMap<String, String>()
+                body.add("grant_type", "authorization_code")
+                body.add("code", code)
+                body.add("redirect_uri", redirectUri)
+                body.add("client_id", clientId)
+                body.add("client_secret", clientSecret)
+                body.add("code_verifier", getOauth2CodeVerifier(session))
+
+                // Exchange code for access token and refresh token
+                val tokenResponse = webClient.post()
+                    .uri(URIBuilder(truidTokenEndpoint).build())
+                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                    .accept(MediaType.APPLICATION_JSON)
+                    .body(fromFormData(body))
+                    .retrieve()
+                    .awaitBody<TokenResponse>()
+
+                // Get and print user email from Truid
+                val getPresentationUri = URIBuilder(truidPresentationEndpoint)
+                    .addParameter("claims", "truid.app/claim/email/v1")
+                    .build()
+
+                val presentation = webClient
+                    .get()
+                    .uri(getPresentationUri)
+                    .accept(MediaType.APPLICATION_JSON)
+                    .header(HttpHeaders.AUTHORIZATION, "Bearer ${tokenResponse.accessToken}")
+                    .retrieve()
+                    .awaitBody<PresentationResponse>()
+
+                println(presentation)
+
+                // Persist token, so it can be accessed via GET "/truid/v1/presentation"
+                // See getAccessToken for an example of refreshing access token
+                persist(tokenResponse)
+            } catch (e: WebClientResponseException.Forbidden) {
+                throw Forbidden("access_denied", e.message)
+            }
+        }
+
+        if (exchange.request.headers.accept.contains(MediaType.TEXT_HTML)) {
+            // Redirect to success page in the webapp flow
+            exchange.response.headers.location = webSuccess
+            exchange.response.statusCode = HttpStatus.FOUND
+            return null
+        } else {
+            // Return a 200 response in case of an AJAX request
+            exchange.response.statusCode = HttpStatus.OK
+            return null
+        }
+    }
+
+    @GetMapping(
+        path = ["/truid/v1/login-presentation"],
+        produces = [MediaType.APPLICATION_JSON_VALUE]
+    )
+    suspend fun getPresentation(): PresentationResponse {
```

Comment:
I'm thinking, maybe we should have an endpoint that is supposed to be authenticated for the user, but does not fetch the presentation from truid.
One of the important points of using the login session is that the user cannot be considered to be logged in if the access token is not valid. This example might very easily give tthe impression that you only need to refresh the token when accessing the truid presentation.

We can have an empty endpoint only containing a comment:

```
@PostMapping()
suspend fun performAction() {
  // Perform action on behalf of the user
}
```

Reply:
Sounds good!